test/integration/cluster_filter_integration_test.cc

```diff
@@ -0,0 +1,124 @@
+#include "envoy/network/filter.h"
+#include "envoy/registry/registry.h"
+
+#include "test/config/utility.h"
+#include "test/integration/integration.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace {
+
+class PoliteFilter : public Network::Filter, Logger::Loggable<Logger::Id::filter> {
+public:
+  Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override {
+    ENVOY_CONN_LOG(debug, "polite: onData {} bytes {} end_stream", read_callbacks_->connection(),
+                   data.length(), end_stream);
+    if (!read_greeted_) {
+      Buffer::OwnedImpl greeter("surely ");
+      read_callbacks_->injectReadDataToFilterChain(greeter, false);
+      read_greeted_ = true;
+    }
+    return Network::FilterStatus::Continue;
+  }
+  Network::FilterStatus onWrite(Buffer::Instance& data, bool end_stream) override {
+    ENVOY_CONN_LOG(debug, "polite: onWrite {} bytes {} end_stream", write_callbacks_->connection(),
+                   data.length(), end_stream);
+    if (!write_greeted_) {
+      Buffer::OwnedImpl greeter("please ");
+      write_callbacks_->injectWriteDataToFilterChain(greeter, false);
+      write_greeted_ = true;
+    }
+    return Network::FilterStatus::Continue;
+  }
+  Network::FilterStatus onNewConnection() override {
+    ENVOY_CONN_LOG(debug, "polite: new connection", read_callbacks_->connection());
+    return Network::FilterStatus::Continue;
+  }
+
+  void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
+    read_callbacks_ = &callbacks;
+  }
+  void initializeWriteFilterCallbacks(Network::WriteFilterCallbacks& callbacks) override {
+    write_callbacks_ = &callbacks;
+  }
+
+private:
+  Network::ReadFilterCallbacks* read_callbacks_{};
+  Network::WriteFilterCallbacks* write_callbacks_{};
+  bool read_greeted_{false};
+  bool write_greeted_{false};
+};
+
+class PoliteFilterConfigFactory
+    : public Server::Configuration::NamedUpstreamNetworkFilterConfigFactory {
+public:
+  Network::FilterFactoryCb
+  createFilterFactoryFromProto(const Protobuf::Message&,
+                               Server::Configuration::CommonFactoryContext&) override {
+    return [](Network::FilterManager& filter_manager) -> void {
+      filter_manager.addFilter(std::make_shared<PoliteFilter>());
+    };
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{new Envoy::ProtobufWkt::Empty()};
```

Comment:
Sorry I should have asked this before but can you add a test in which there is some basic config for this filter? Also nit, this can be make_unique, possibly same elsewhere.

Reply:
Done